### ISSUE TITLE
class type inference error

### ISSUE THREAD

**Issue by chenkovsky** [crystal-lang/crystal#2589]
the following code cannot be compiled in 0.1.6.
it complains " wrong number of arguments for 'A.new'".
```cr
class A
  def initialize(@a : Int32, @b : Int32)
  end
end

class B < A
  def initialize(@a : Int32)
    super(@a, 1)
  end
end

class C < A
  def initialize(@a : Int32)
    super(@a, 1)
  end
end

class D < A
end

class E
  getter :e

  def initialize(@e : (B.class | C.class))
  end
end

e = E.new B

e.e.new 1 # it compiles ,if you comment this line.

class F < B
end

class G < B
end

class H
  getter :h

  def initialize(@h : (F.class | G.class))
  end
end

h = H.new F
h.h.new 1 # this is correct

`

**Comment by Nephos** [comment]
don't you have to call super() to use the parent constructor ?

**Comment by Nephos** [comment]
It's "funny", B.new 1 work, and the constructors calls seems good (B, A via super with the right arguments)

**Comment by asterite** [comment]
When you do `B.class | C.class` the compiler will actually type it as `A+:Class`, a virtual metaclass. This is described here: http://crystal-lang.org/docs/syntax_and_semantics/virtual_and_abstract_types.html (I think it might only be described for instances, but it also applies for metaclasses). So, the compiler analyzes all potential `A.new`, `B.new` and `C.new` calls with those arguments, but `A.new` with a single argument doesn't exist.

This might be improved in the future (we'll probably don't do this type merge automatically).

If `A` is never supposed to be initialized you can mark it as abstact and it will work... mmm... actually, no, there's also `D` which inherits `A` and has two arguments.

What are the real classes behind A, B, C, D...? Maybe it can be modelled in another way (or the current way isn't right)

**Comment by ozra** [comment]
Is there an issue for A|B => Ancestor? This is really annoying in many cases actually, I don't really see the benefit, but many hassles.

**Comment by asterite** [comment]
The automatic merge behaviour was mainly to avoid having several "similar" unions. For example the compiler has a huge ASTNode hierarchy. If we didn't do this we'd have `Var | Call`, `Var | Splat`, `Call | Splat`, etc., instead of just `ASTNode+` in many places.

This was specially true before 0.16.0:

```crystal
class Splat
  def initialize(@exp)
  end
end

Splat.new(Call.new) # @exp is now Call
Splat.new(Var.new) # @exp is now Call | Var
Splat.new(NumberLiteral.new(1)) # @exp is now Call | Var | NumberLiteral
```

Each time the type changes, potentially a lot of code needs to be re-typed. You can imagine this to be very slow.

However, and because now instance variables types must be specified, this becomes:

```crystal
class Splat
  # @exp is now ASTNode (well, ASTNode+, the compiler does this without you noticing it)
  # and the type can never change
  def initialize(@exp : ASTNode)
  end
end

Splat.new(Call.new) # @exp is still ASTNode+, merging ASTNode+ with Call is ASTNode+
Splat.new(Var.new) # same...
```

So now that the global type inference changed we can improve this too :-)

I'll probably try to implement it after 0.17.0, right now I'm more focused on method arguments, splats, double splats, etc.

**Comment by ozra** [comment]
Ah, neat! :-)

**Comment by jhass** [comment]
What's the status here? Why is this issue open with no tags? Is it a compiler bug? Is it a duplicate of some other issue? Is it a question, if so answered?

**Comment by asterite** [comment]
We should probably create a separate issue that proposes not merging types under the same hierarchy as a base "virtual" type.

**Comment by jhass** [comment]
Could you do that please and close here then?

**Comment by asterite** [comment]
Closed in favor of #2661
